### DISCUSSION TITLE
[Linux / Wayland / OTD] Tablet movement not detected at all

### DISCUSSION THREAD

**Discussion by fxzzi** [ppy/osu#30625] [Q&A]
![image](https://github.com/user-attachments/assets/0d0133d4-d36b-4ed7-bf88-692a1b4b2e89)

i have my tablet mapped like so. This seems to work perfectly when not in osu!. Whilst within osu!, the mouse cursor in game does not respond to any pen movements, whilst my mouse still moves about fine. The game does respond to the pen clicks (as in pressing the surface of the tablet with the pen will trigger a left click in game as expected).

I have tablet disabled within the osu!lazer settings. If I enable it and setup my tablet within osu!, it then seems to function just fine. Is this intended behaviour? Am I supposed to use the in game tablet set up instead of osu! detecting any external tablet software?

**Comment by SupDos** [comment]
As per the [OTD Wiki](https://opentabletdriver.net/Wiki/FAQ/LinuxAppSpecific#osu-lazer-broken-input-wayland) you need to set `SDL_VIDEODRIVER` to `wayland`

```
env SDL_VIDEODRIVER=wayland ./osu.AppImage
```

If that still doesn't work though probably best to ask in their discord server for help.

**Reply by fxzzi** [reply]
> As per the [OTD Wiki](https://opentabletdriver.net/Wiki/FAQ/LinuxAppSpecific#osu-lazer-broken-input-wayland) you need to set `SDL_VIDEODRIVER` to `wayland`
> 
> ```
> env SDL_VIDEODRIVER=wayland ./osu.AppImage
> ```
> 
> If that still doesn't work though probably best to ask in their discord server for help.

thanks for your reply. I have however already added this env var and confirmed it is running with wayland native. `hyprctl clients` shows `xwayland: 0` for the osu! window. I will ask in their discord and see if the issue is with them or with osu!. Considering that otd seems to be working fine for me with everything but osu!, i assumed it was an issue on that front.

**Comment by Stoppedpuma** [comment]
See if switching to artist mode solves the issue.

**Reply by fxzzi** [reply]
> See if switching to artist mode solves the issue.

artist mode seems to make the cursor move in game again, however the cursor is stuck on the right side of the screen and there is also some smoothing

**Reply by Stoppedpuma** [reply]
Disable raw input and add the following lines to `/etc/libinput/local-overrides.quirks`:

[OpenTabletDriver Virtual Tablet]
MatchName=OpenTabletDriver*
AttrTabletSmoothing=0

If disabling raw input doesn't work then it's a hyprland bug. I had the same issue when I used hyprland and I believe the work around was to map both displays in OTD and set the tablet output to entire in hyprland.conf?

**Reply by fxzzi** [reply]
> Disable raw input and add the following lines to `/etc/libinput/local-overrides.quirks`:
> 
> [OpenTabletDriver Virtual Tablet] MatchName=OpenTabletDriver* AttrTabletSmoothing=0
> 
> If disabling raw input doesn't work then it's a hyprland bug. I had the same issue when I used hyprland and I believe the work around was to map both displays in OTD and set the tablet in output to entire in hyprland.conf?

thanks for the explanation on how to disable the smoothing. However raw input is already disabled. The Hyprland bug you mentioned was fixed a few days ago: https://github.com/hyprwm/Hyprland/pull/8352

as I said before, it works perfectly on everything but osu, and even works if I am focused on my second screen but moving my pen around on the unfocused osu! window (i can tell its unfocused, the fps is limited), clicking on the osu! window to focus it makes the problem appear again.

**Reply by Stoppedpuma** [reply]
The previous PR which was supposed to "fix" this didn't work either (at least in osu!s case), the underlying problem here is still going to be Hyprland as other wayland compositors do not face this issue. I would open a bug report on Hyprlands side and see if you can find a work around in the mean time.

**Reply by fxzzi** [reply]
> The previous PR which was supposed to "fix" this didn't work either (at least in osu!s case), the underlying problem here is still going to be Hyprland as other wayland compositors do not face this issue. I would open a bug report on Hyprlands side and see if you can find a work around in the mean time.

I did have the exact issue you spoke of on all of my desktop previously, however that was fixed in Hyprland with that PR I mentioned earlier. If its fixed in all of the desktop, how could it possibly be Hyprlands doing for artist mode to still be broken, and only within osu!? My apologies for being naive, I am not completely sure how the tablet works within lazer / otd / the compositor.

**Reply by Stoppedpuma** [reply]
The artist mode for lazer on wayland problem has been an issue on OpenTabletDrivers end for a few years now. The movement part about it working fine outside of osu! applied to this PR as well https://github.com/hyprwm/Hyprland/pull/7586. There tends to be a pattern of x or y program working perfectly on everything except Hyprland, this would be such as case ever since Hyprland changed how tablets work ~6 months ago.

This boils down to being Hyprlands issue as osu! works fine with tablets on every other wayland compositor in my testing (KDE, Gnome, Sway, River). If you would like to confirm this, I suggest giving Sway a try.

**Reply by fxzzi** [reply]
thanks for the explanation. how come we have to use artist mode in the first place? Shouldn't absolute mode just... work fine? However it doesn't. Whose side is that on, Hyprland also?

**Reply by Stoppedpuma** [reply]
That would be on OpenTabletDrivers side, I couldn't tell you why exactly since I haven't looked into the code but IIRC there was a PR which fixed it along the way on `master`?

**Comment by fxzzi** [comment]
sorry for all the confusion, looks like absolute mode is indeed working now. Even though as I said earlier, high precision mouse was already disabled, i toggled it on and back off again and it seems to be working. Now I just have one remaining issue, i think this stemmed from the SDL3 move. My game launches in the correct resolution (2560x1440), however if I receive a notification my game drops to 1080p, which is the resolution of my second monitor. I can fix this by changing my monitor in osu! settings to the correct one, but this setting seems to reset every time I relaunch osu!: 
![image](https://github.com/user-attachments/assets/0af4fd28-c081-4602-966c-1409e16864dd)

**Reply by Stoppedpuma** [reply]
This seems to happen under Sway as well, it shouldn't be SDL3 specific as it occurs under SDL2. I haven't had this happen under SDL3 yet though so I'll report back on this eventually.
